### PR TITLE
antispoof: allow more than one ipv4 address

### PR DESCRIPTION
**Pull request by outscale-jju**
Signed-off-by: Jerome Jutteau <jerome.jutteau@outscale.com>

### CODE REVIEW

File: src/antispoof.c [hunk 1]
```diff
@@ -24,6 +24,8 @@
 #include "utils/bitmask.h"
 #include "utils/mac.h"
 
+#define ARP_MAX 100
+
 struct pg_antispoof_arp {
```
Comment (outscale-mgo) [inline src/antispoof.c]: I'm not sure about the alignment of this stuct, maybe we can

File: src/antispoof.c [hunk 2]
```diff
@@ -24,6 +24,8 @@
 #include "utils/bitmask.h"
 #include "utils/mac.h"
 
+#define ARP_MAX 100
```
Comment (outscale-mgo) [inline src/antispoof.c]: can you use a power of 2 here

File: src/antispoof.c [hunk 3]
```diff
@@ -64,23 +72,77 @@ struct pg_antispoof_config {
...
-	state = pg_brick_get_state(brick, struct pg_antispoof_state);
-	state->ip = ip;
+int pg_antispoof_arp_add(struct pg_brick *brick, uint32_t ip,
+			 struct pg_error **errp)
+{
+	struct pg_antispoof_state *state =
+		pg_brick_get_state(brick, struct pg_antispoof_state);
+	uint16_t n = state->arp_size;
+
+	if (n == ARP_MAX) {
```
Comment (outscale-mgo) [inline src/antispoof.c]: unlikely

File: src/antispoof.c [hunk 4]
```diff
@@ -64,23 +72,77 @@ struct pg_antispoof_config {
...
+		*errp = pg_error_new("Maximal IP reached");
+		return -1;
+	}
+	struct arp *arp = &state->arp[n];
```
Comment (outscale-mgo) [inline src/antispoof.c]: you can do this before the check, as long as you don't access arp, you should not get any segfault

File: src/antispoof.c [hunk 5]
```diff
@@ -45,12 +47,18 @@ struct pg_antispoof_arp {
...
+
 struct pg_antispoof_state {
 	struct pg_brick brick;
 	enum pg_side outside;
 	struct ether_addr mac;
-	uint32_t ip;
-	struct pg_antispoof_arp arp_packet;
+	bool arp_enabled;
+	uint16_t arp_size;
+	struct arp arp[ARP_MAX];
```
Comment (outscale-mgo) [inline src/antispoof.c]: s/struct arp arp/struct arp arps/

File: src/antispoof.c [hunk 6]
```diff
@@ -101,13 +163,20 @@ static struct pg_brick_config *antispoof_config_new(const char *name,
...
-		return 0;
+	uint16_t size = state->arp_size;
+	struct pg_antispoof_arp *p;
+
+	for (uint16_t i = 0; i < size; i++) {
+		/* Check that all fields match reference packet except the
+		 * "ar_op" (arp operation code) part.
+		 */
+		p = &state->arp[i].packet;
+		if (memcmp(p, a, PG_ANTISPOOF_ARP_CHECK_PART1) == 0 &&
```
Comment (outscale-mgo) [inline src/antispoof.c]: I would use ! instead of `== 0`

File: src/antispoof.c [hunk 7]
```diff
@@ -144,21 +213,22 @@ static int antispoof_burst(struct pg_brick *brick, enum pg_side from,
...
 
 		/* ARP antispoof */
-		if ((eth->ether_type == htobe16(ETHER_TYPE_ARP)) && state->ip) {
-			struct pg_antispoof_arp *a;
-
-			a = (struct pg_antispoof_arp *)(eth + 1);
-			if (antispoof_arp(state, a) < 0) {
+		if (state->arp_enabled) {
+			if (eth->ether_type ==
+			    rte_cpu_to_be_16(ETHER_TYPE_ARP)) {
```
Comment (outscale-mgo) [inline src/antispoof.c]: can't we do the rte_cpu_to_be_16 at compille time ?

File: src/antispoof.c [hunk 8]
```diff
@@ -144,21 +213,22 @@ static int antispoof_burst(struct pg_brick *brick, enum pg_side from,
 		}
 
 		/* ARP antispoof */
-		if ((eth->ether_type == htobe16(ETHER_TYPE_ARP)) && state->ip) {
-			struct pg_antispoof_arp *a;
-
-			a = (struct pg_antispoof_arp *)(eth + 1);
-			if (antispoof_arp(state, a) < 0) {
+		if (state->arp_enabled) {
+			if (eth->ether_type ==
```
Comment (outscale-mgo) [inline src/antispoof.c]: can we have something like
```
if ((eth->ether_type == rte_cpu_to_be_16(ETHER_TYPE_ARP) && antispoof_arp(state, eth + 1) < 0) ||
    eth->ether_type == rte_cpu_to_be_16(ETHER_TYPE_RARP))) {
```

 also, you could use a temporary variable for eth->ether_type, and I would use unlikely on this condition, as most packets won't be arp, it's a good deal to accelerate all non arp packets

File: src/antispoof.c [hunk 9]
```diff
@@ -144,21 +213,22 @@ static int antispoof_burst(struct pg_brick *brick, enum pg_side from,
...
+			    rte_cpu_to_be_16(ETHER_TYPE_ARP)) {
+				struct pg_antispoof_arp *a;
+
+				a = (struct pg_antispoof_arp *)(eth + 1);
+				if (antispoof_arp(state, a) < 0) {
+					pkts_mask &= ~bit;
+					continue;
+				}
+			} else if (eth->ether_type ==
```
Comment (outscale-mgo) [inline src/antispoof.c]: I guess at the end you should have a condition similar to:
```
if ((eth_type == BE_ETHER_TYPE_ARP) && antispoof_arp(state, eth + 1) < 0) ||
    eth_type == BE_ETHER_TYPE_RARP))) {
```

**Comment by outscale-jju** [comment]
merged

**Pull request closed without merging**